rfcs/20240808-graph-api-int-compression-for-sdpa/README.md

```diff
@@ -0,0 +1,307 @@
+`per_tensor` or `per_channel`. And `axis` specifies on which dimension
+`per_channel` quantization will be applied. To support int4 data types,
+`DynamicQuantize` and `DynamicDequantize` operations in oneDNN Graph opset will
+be extended:
+
+1. `DynamicQuantize` will be extended to support u4/s4 output, and scales and
+   zero points with groups. The library may raise an error if groups are not
+   provided for a `DynamicQuantize` op with u4/s4 output.
+2. `DynamicDequantize` will be extended to support u4/s4 input, and scales and
+   zero points with groups. The library may raise an error if groups are not
+   provided for a `DynamicDequantize` op with u4/s4 input.
+
+Besides the input or output data types, new attributes will be added for
+DynamicQuantize and DynamicDequantize:
+
+1. Add `per_group` to the supported values of `qtype` attribute, and the default value
+   will be unchanged.
+2. Add a new optional attribute `group_size` to support `per_group` quantization
+   type. The value of `group_size` attribute is a single `s64` data, which
+   indicates the number of elements that share the same scaling factor and zero
+   points in each quantization group. The attribute is only required when
+   `per_group` quantization type is specified for `qtype`. If `per_group`
+   quantization is not specified and `group_size` attribute is given, it will
+   be ignored. 
+3. The `scale` and `zp` input will be extended to a `n`-dimensional tensor
+   (`n` = dims(input)) for `per_group` quantization. The shape requirements for
+   `scale` and `zp` are as follows:
+   1. For `per_tensor` and `per_channel` quantization, the `scale` and `zp`
+      inputs should be `1d` tensors.
+   2. For `per_group` quantization, all dimensions should match the input,
+      excepts for the dimension where grouped quantization applies, which should
+      be `src_dim / group_size`.
+
+
+For example, if the shape of Key tensor is `4x8x4`, and the group size is `4`,
+which means each scaling factor will be adopted for 4 times. One possible scale
+would be:
+
+![alt text](img/grouped_scale.png)
+
+The usage of the new attribute will be like:
+
+```cpp
+using namespace dnnl::graph;
+
+const dim K = 10, N = 20;
+const size_t G = N / 2;
+
+dims src_dims = {K, N};
+dims scale_dims = {K, N/G};
+
+// In real case, users do not have to provide the concrete shapes of logical
+// tensors until compilation stage. 
+logical_tensor src_desc {0, data_type::u4, src_dims, layout_type::strided};
+logical_tensor scale_desc {
+        1, data_type::f32, scale_dims, layout_type::strided};
+logical_tensor dst_desc {2, data_type::f32, src_dims, layout_type::strided};
+
+op deq = op(0, op::kind::DynamicDequantize, {src_desc, scale_desc},
+        {dst_desc}, "deq");
+// Specify the quantization type as per_group quantization.
+deq.set_attr<std::string>(op::attr::qtype, "per_group");
+// Group size indicates the the size of each group.
+deq.set_attr<int64_t>(op::attr::group_size, G);
```

Comment:
Is `axis` for a group implied here?

Reply:
Thanks for reminding. We have updated the `axis` as an optional attribute for per-group quantization.